### DISCUSSION TITLE
Does blink.cmp support buffer local sources?

### DISCUSSION THREAD

**Discussion by gh-liu** [saghen/blink.cmp#361] [General]
Does blink.cmp support buffer local sources? kind of like `cmp.setup.filetype()` which set sources for specified type.

**Comment by saghen** [comment]
Hey there, providers will be chosen before the menu shows up via `enabled_providers` which can be a function. This also works for i.e. choosing providers based on the current treesitter node

```lua
sources = {
  completion = {
    enabled_providers = function()
      local node = vim.treesitter.get_node()
      if vim.bo.filetype == 'toml' then
        return { 'crates' }
      elseif node and vim.tbl_contains({ 'comment', 'line_comment', 'block_comment' }), node:type())
        return { 'buffer' }
      else
        return { 'lsp', 'buffer', etc.. }
      end
    end
  }
}
```

**Comment by gh-liu** [comment]
Thanks for your quick reply, it works.
By the way, this plugin is awsome.
